Feature/cxspa 206

kimhw0630 opened the pull request.

hakwookim committed:
first commit for ASM

Co-authored-by: Eric Polesky <eric@unboundcommerce.com>

hakwookim committed:
Co-authored-by: Eric Polesky <eric@unboundcommerce.com>

hakwookim committed:
Fiori

ercultimate committed:
Merge remote-tracking branch 'origin/develop' into feature/cxspa-206

ercultimate committed:
implement asm query service

ercultimate committed:
add customer search to asm facade

- move more of model into root to support new facade intnerface

hakwookim committed:
remove fiori and ngrx for customer group lists

hakwookim committed:
test sort

hakwookim committed:
update design

ercultimate committed:
use modified query with parameters and include additional ui updates

ercultimate committed:
add beta parameter support to query system

hakwookim committed:
pagenation

hakwookim committed:
fix unit tests

hakwookim committed:
added spinner while waiting for backend

hakwookim committed:
unit tests for customer list & main ui components

hakwookim committed:
some style changes

hakwookim committed:
missing localization

hakwookim committed:
e2e basic

hakwookim committed:
updated sorting to handle dynamically

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

hakwookim committed:
remove unused property

hakwookim committed:
rollback yarn lock file

hakwookim committed:
update peerDependencies

hakwookim committed:
fix for css property-disallowed-list

hakwookim committed:
re-generate dependencies.json

ercultimate committed:
fix asm/root circular dependency with self

hakwookim committed:
UX change for list

hakwookim committed:
column title change for email

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

ercultimate committed:
Revert "add beta parameter support to query system"

This reverts commit 3e61c09b7859f927c289d7b8e3ccf5a81992d99a.

ercultimate committed:
remove beta query usage

hakwookim committed:
response design

hakwookim committed:
update breakpoint

hakwookim committed:
update unit tests

hakwookim committed:
fix prettier format

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

ercultimate committed:
add unit tests for asm-query service

giancorderoortiz committed:
Merge branch 'develop' into feature/cxspa-206

hakwookim committed:
accessibility table cell

hakwookim committed:
use cxFocus

hakwookim committed:
updated e2e test

hakwookim committed:
configuration for asm customer list

Co-authored-by: Eric Polesky <ercultimate@users.noreply.github.com>

hakwookim committed:
fix build issue with using protected property

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

hakwookim committed:
fix avatar position

hakwookim committed:
fix accessibility for ng-select

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

hakwookim committed:
fix build error

hakwookim committed:
replaced with cx-sort

hakwookim committed:
fix accessibility issue with dialog

hakwookim committed:
fix unit test for asm main

hakwookim committed:
rename style class name to follow cx-naming

hakwookim committed:
fix accessibility issue in sort

morganm58 commented on feature-libs/asm/styles/components/_asm-main-ui.component.scss:
```diff
@@ -103,6 +103,20 @@
+      .cx-asm-customer-list {
+        cursor: pointer;
+        cx-icon {
+          padding: 0 7px 0 0;
```
for directionality this one should be
```
padding-top: 0;
padding-inline-end: 7px;
padding-bottom: 0;
padding-inline-start: 0;
```

morganm58 commented on feature-libs/asm/styles/components/_customer-list.component.scss:
```diff
@@ -0,0 +1,185 @@
+          color: $sapFioriLinkColor;
+        }
+        cx-icon.previous {
+          padding: 0 10px 0 0;
```
This one should be changed for directionality as well.
```
padding-top: 0;
padding-inline-end: 10px;
padding-bottom: 0;
padding-inline-start: 0;
```

morganm58 commented on feature-libs/asm/styles/components/_customer-list.component.scss:
```diff
@@ -0,0 +1,185 @@
+          padding: 0 10px 0 0;
+        }
+        cx-icon.next {
+          padding: 0 0 0 10px;
```
This one should be changed for directionality as well.
```
padding-top: 0;
padding-inline-end: 0;
padding-bottom: 0;
padding-inline-start: 10px;
```

morganm58 commented on feature-libs/asm/styles/components/_customer-list.component.scss:
```diff
@@ -0,0 +1,185 @@
+        }
+        .cx-btn-previous {
+          span {
+            padding: 0 0 0 5px;
```
This one should be changed for directionality as well.
```
padding-top: 0;
padding-inline-end: 0;
padding-bottom: 0;
padding-inline-start: 5px;
```

morganm58 commented on feature-libs/asm/styles/components/_customer-list.component.scss:
```diff
@@ -0,0 +1,185 @@
+        }
+        .cx-btn-next {
+          span {
+            padding: 0 5px 0 0;
```
This one should be changed for directionality as well.
```
padding-top: 0;
padding-inline-end: 5px;
padding-bottom: 0;
padding-inline-start: 0;
```

morganm58 commented on feature-libs/asm/styles/components/_customer-list.component.scss:
```diff
@@ -0,0 +1,185 @@
+        div:first-child {
+          font-size: 0.875rem;
+          color: $sapFioriTextColor;
+          padding: 0 10px 0 10px;
```
this one can be changed to 
padding: 0 10px;

morganm58 commented on feature-libs/asm/styles/components/_customer-list.component.scss:
```diff
@@ -0,0 +1,185 @@
+      }
+      .table-header-cell {
+        line-height: 40px;
+        margin: 0 15px 0 0;
```
This one should be changed for directionality as well.
```
margin-top: 0;
margin-inline-end: 15px;
margin-bottom: 0;
margin-inline-start: 0;
```

morganm58 commented on feature-libs/asm/styles/components/_customer-list.component.scss:
```diff
@@ -0,0 +1,185 @@
+      text-align: center;
+      background: #9259cd;
+      text-transform: uppercase;
+      margin: 2px 10px 2px 0;
```
This one should be changed for directionality as well.
```
margin-top: 2px;
margin-inline-end: 10px;
margin-bottom: 2px;
margin-inline-start: 0;
```

morganm58 commented on feature-libs/asm/styles/components/_customer-list.component.scss:
```diff
@@ -0,0 +1,185 @@
+  }
+  ng-select {
+    .ng-select-container {
+      border: 0 !important;
```
We try to stay away from using !important in our scss. can this be made more specific to avoid the use of the !important?

morganm58 commented on feature-libs/asm/styles/components/_customer-list.component.scss:
```diff
@@ -0,0 +1,185 @@
+      border: 0 !important;
+      .ng-arrow-wrapper {
+        .ng-arrow {
+          border-color: $sapFioriLinkColor transparent transparent !important;
```
Same comment as above. Can this be made more specific to avoid the use of the !important?

morganm58 commented on feature-libs/asm/styles/components/_customer-list.component.scss:
```diff
@@ -0,0 +1,185 @@
+    }
+  }
+  .ng-select-focused {
+    outline: none !important;
```
Same comment as above. Can this be made more specific to avoid the use of the !important.

hakwookim committed:
fix for style code review

hakwookim committed:
code review style fix

morganm58 commented:
The pagination doesn't match what is in the visual. The dropdown styling is also different
<img width="283" alt="Screen Shot 2022-07-01 at 11 05 16 AM" src="https://user-images.githubusercontent.com/2947470/176921737-e5a835c5-15ce-4b98-bcbd-862650135abb.png">
<img width="398" alt="Screen Shot 2022-07-01 at 11 05 23 AM" src="https://user-images.githubusercontent.com/2947470/176921741-20c32aa5-898f-42c6-9512-cd763d125be5.png">
![Screen Shot 2022-07-01 at 11 05 33 AM](https://user-images.githubusercontent.com/2947470/176921743-1109a771-6cf4-4022-b934-944bf5a5cc9d.png)
<img width="1606" alt="Screen Shot 2022-07-01 at 11 12 10 AM" src="https://user-images.githubusercontent.com/2947470/176921745-608164b8-5a0f-4aed-8e46-fcf943fc6590.png">

kimhw0630 commented:
> The pagination doesn't match what is in the visual. The dropdown styling is also different <img alt="Screen Shot 2022-07-01 at 11 05 16 AM" width="283" src="https://user-images.githubusercontent.com/2947470/176921737-e5a835c5-15ce-4b98-bcbd-862650135abb.png"> <img alt="Screen Shot 2022-07-01 at 11 05 23 AM" width="398" src="https://user-images.githubusercontent.com/2947470/176921741-20c32aa5-898f-42c6-9512-cd763d125be5.png"> ![Screen Shot 2022-07-01 at 11 05 33 AM](https://user-images.githubusercontent.com/2947470/176921743-1109a771-6cf4-4022-b934-944bf5a5cc9d.png) <img alt="Screen Shot 2022-07-01 at 11 12 10 AM" width="1606" src="https://user-images.githubusercontent.com/2947470/176921745-608164b8-5a0f-4aed-8e46-fcf943fc6590.png">

@morganm58 
For the drop-down, we are using common controls (ng-select and cx-sort).
For the pagination currently we don't get all required fields from the backend so we can't match the design right now.  We have already informed Veronica.

hakwookim committed:
Merge remote-tracking branch 'origin/develop' into feature/cxspa-206

hakwookim committed:
added feature flag

hakwookim committed:
update unit test

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

WeizhengSap commented on feature-libs/asm/root/facade/asm-facade.service.ts:
```diff
@@ -0,0 +1,24 @@
+import { Injectable } from '@angular/core';
+import { facadeFactory, QueryState } from '@spartacus/core';
```
can we rename it to `customer-list.facade.ts`? Or merge these 2 functions into `asm.facade.ts` which is created in https://github.com/SAP/spartacus/pull/15776/files.

WeizhengSap commented on feature-libs/asm/root/index.ts:
```diff
@@ -0,0 +1,2 @@
+export * from './facade/index';
+export * from './model/index';
```
We can put all exported stuff in index.ts
and `export * from './index';` in public_api.ts.

WeizhengSap commented on feature-libs/asm/root/model/customer-lists-page.model.ts:
```diff
@@ -0,0 +1,14 @@
+import { UserGroup } from './user-group.model';
```
can we put these model interfaces into one file something like this: `customer-list.model.ts`

WeizhengSap commented on feature-libs/asm/occ/adapters/occ-asm.adapter.ts:
```diff
@@ -40,9 +69,16 @@ export class OccAsmAdapter implements AsmAdapter {
+      params = params.set('sort', '' + options.sort);
+    } else {
+      params = params.set('sort', 'byNameAsc');
+    }
```
can be this `params = params.set('sort', options.sort ?? 'byNameAsc');`.

ercultimate replied:
This suggestion is not sufficient.  There are 3 cases here:
1. Sort is provided
   - use provided value 
1. Sort not provided, normal customer search
   - use default value
1. Sort not provided, customer list search
   - do not send default, let api select default value
  
The valid sort values for a customer list search are defined per the search strategy used for the list.  The sort value for a normal customer search is highly likely to not be valid.  Unless it is explicitly specified, the sort value is not sent so that the API can respond with its default sort method.  The chosen sort value is readable in the response should you need to make UI adjustments.

As a side note, I don't believe we should be providing a hard-coded default for sort for any case.

WeizhengSap commented on feature-libs/asm/occ/adapters/occ-asm.adapter.ts:
```diff
@@ -52,6 +88,14 @@ export class OccAsmAdapter implements AsmAdapter {
     }
 
+    if (typeof options['currentPage'] !== 'undefined') {
+      params = params.set('currentPage', '' + options.currentPage);
```
why `'' + options.currentPage`? convert number to string? can it be this?: 
```
if (options.pageSize !== undefined) {
      params = params.set('pageSize', options.pageSize);
    }
```

WeizhengSap commented on feature-libs/asm/occ/adapters/occ-asm.adapter.ts:
```diff
@@ -29,7 +33,32 @@ export class OccAsmAdapter implements AsmAdapter {
+  }
+
+  customerLists(): Observable<CustomerListsPage> {
+    const headers = InterceptorUtil.createHeader(
```
this header seems used in all functions in this adapter. Maybe make it as `protected` class variable.

WeizhengSap commented on feature-libs/asm/core/facade/asm.service.ts:
```diff
@@ -14,7 +17,20 @@ import { AsmSelectors } from '../store/index';
 })
 export class AsmService {
-  constructor(protected store: Store<StateWithAsm>) {}
+  constructor(
```
facade codes need to be organized with the code in this PR https://github.com/SAP/spartacus/pull/15776/files. Either create 2 facades `asm facade` and `customers facade`, or just one facade`asm facade`.

WeizhengSap commented on feature-libs/asm/core/facade/asm-query.service.ts:
```diff
@@ -0,0 +1,53 @@
+import { AsmConnector } from '../connectors/asm.connector';
+
+@Injectable()
+export class AsmQueryService implements AsmFacadeService {
```
`AsmQueryService` is not a good name. We can just create queries in facade service.

WeizhengSap commented on feature-libs/asm/core/facade/asm-query.service.ts:
```diff
@@ -0,0 +1,53 @@
+  getCustomers(
+    options: CustomerSearchOptions = {}
+  ): Observable<CustomerSearchPage> {
+    return this.customerSearchCommand$.execute(options);
```
I understand we use `command` because we cannot pass parameter to `query`. Can you take a look of this code, and consider whether we can use a similar way here.
```
protected searchInput$ = new Subject<{ term: string; pagination: number, ... }>();
performSearch(term: string, pagination: number, ...): void {
  this.searchInput$.next({ term, pagination, ... });
}
protected searchQuery: Query<unknown> = this.query.create(
  () =>
    this.searchInput$.pipe(
      switchMap(({ term, pagination, ... }) =>
        this.connector.search(term, pagination, ...)
      )
    )
);
```

ercultimate replied:
Can you explain more why the suggestion is more desirable?  

As I understand, the suggestion is what the Command object does.  The critical behavioral difference is that, with the suggestion, all consumers of `getCustomers()` would receive the same value.  The connector result acts like a "singleton" provider, where there can only be 1 value that is the latest page request.  This would prevents multiple consumers from having separate page states.

WeizhengSap commented on feature-libs/asm/components/customer-list/customer-list.model.ts:
```diff
@@ -0,0 +1,7 @@
+import { CustomerListColumnActionType } from '@spartacus/asm/root';
+import { User } from '@spartacus/core';
+
+export interface CustomerListActionEvent {
```
Can we rename to it `CustomerListAction`? We have `CxEvent`, this interface is not Spartacus' event.

WeizhengSap commented on feature-libs/asm/components/customer-list/customer-list.component.ts:
```diff
@@ -0,0 +1,243 @@
+    this.customerListsPage$ = this.asmService
+      .getCustomerLists()
+      .pipe(
+        filter((queryState) => queryState.loading === false),
```
can we move this `filter` and `map` into service? So, calling `this.asmService.getCustomerLists()` just get observable of customerListsPage.

WeizhengSap commented on feature-libs/asm/components/customer-list/customer-list.component.ts:
```diff
@@ -0,0 +1,243 @@
+        })
+      )
+      .pipe(
+        map((customerListsPage) => customerListsPage),
```
why we have this `map` and `filter`?

WeizhengSap commented on feature-libs/asm/components/customer-list/customer-list.component.ts:
```diff
@@ -0,0 +1,243 @@
+  }
+
+  ngOnInit(): void {
+    this.customerListsPage$ = this.asmService
```
we can move this logic out of `ngOnInit`. So, just `customerListsPage$: Observable<CustomerListsPage | undefined> = this.asmService.getCustomerLists()....`

ercultimate replied:
I would prefer not to implement this change. This can make the component more involved to unit test.  

Assume the component initialization is done during class construction, per the suggestion.  In the unit test, the component class is typically already constructed by the time a specific unit test or describe block becomes active.  If you need to change the return values of spies for that test/describe, you need to re-construct the component so that the new values are used in the component initialization.

WeizhengSap commented on feature-libs/asm/components/customer-list/customer-list.component.ts:
```diff
@@ -0,0 +1,243 @@
+        pageSize: pageSize,
+        currentPage: this.currentPage,
+      };
+      if (this.sortCode) {
```
if `this.sortCode` is empty string `''`, it is also false. So, each time, we select a customer group, the sort is always `byNameAsc`

kimhw0630 replied:
The sorting depends on the customer group for example if the customer group id is 'A' then we only can sort by 'date' and if the customer group id is 'B' then we only can sort by 'name'.  If wrong sorting is passed then the backend ignores and send back us the default sorting.  so we removed the default sort (byNameAsc) if the customer group is passed in occ-asm.adapter.ts

WeizhengSap commented on feature-libs/asm/components/customer-list/customer-list.component.ts:
```diff
@@ -0,0 +1,243 @@
+  onChangeCustomerGroup(): void {
+    this.currentPage = 0;
+    this.sorts = null;
+    this.sortCode = '';
```
we set `this.sortCode` to empty string, but in line 103, it will be false.

WeizhengSap commented on feature-libs/asm/components/customer-list/customer-list.component.ts:
```diff
@@ -0,0 +1,243 @@
+      if (this.sortCode) {
+        options.sort = this.sortCode;
+      }
+      this.customerSearchPage$ = of(options)
```
why we use `of` and `switchMap` here? can it be just this?:
```
    const options: CustomerSearchOptions = {
        customerListId: this.selectedUserGroupId,
        pageSize: this.customerListConfig?.pageSize ?? this.PAGE_SIZE,
        currentPage: this.currentPage,
        sort: this.sortCode,
      };

      this.loaded = false;
      this.customerSearchPage$ = this.asmService.searchCustomers(options).pipe(
        tap((result) => {
          this.loaded = true;
          ....
        })
      );
```

WeizhengSap commented on feature-libs/asm/components/customer-list/customer-list.component.ts:
```diff
@@ -0,0 +1,243 @@
+              this.sortCode = result.pagination?.sort;
+            }
+            if (result.entries.length < pageSize) {
+              this.maxPage = result.pagination?.currentPage || 0;
```
`maxPage` here is confusing. Maybe replaced by a boolean variable `isLastPage`. (`this.isLastPage = result.entries.length < pageSize`)

hakwookim committed:
change for code review

ercultimate committed:
remove breaking changes from customer list feature

- move migrated models back to original location
- remove facade methods to prevent cyclical dependency cuased by
  depending on asm/core in asm/root module
- implement replacement ngrx state and selectors for removed facade method

ercultimate committed:
rename asm-facade and asm-query service to more appropriate values

ercultimate committed:
remove barrel from root of asm/root module

ercultimate committed:
update import paths for unmigrated models

ercultimate committed:
fix switched cypress helper interceptor aliases

ercultimate committed:
fix misordered constructor arguments

hakwookim committed:
clean up code

ercultimate committed:
update unit tests for ngrx customer list customers search

hakwookim committed:
export a service

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

hakwookim committed:
fix lint error

ercultimate committed:
fix breaking change in asm service constructor

hakwookim committed:
Add missing license header to sources

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

hakwookim committed:
Merge remote-tracking branch 'origin/develop' into feature/cxspa-206

hakwookim committed:
update public_api.ts for customer list

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

WeizhengSap commented on feature-libs/asm/occ/adapters/occ-asm.adapter.ts:
```diff
@@ -35,27 +37,69 @@ export class OccAsmAdapter implements AsmAdapter {
+
+    return this.http
+      .get<CustomerListsPage>(url, { headers, params })
+      .pipe(this.converterService.pipeable(CUSTOMER_LISTS_NORMALIZER));
```
better handle error as well: `catchError((error) => throwError(normalizeHttpError(error))),`

WeizhengSap commented on feature-libs/asm/core/store/actions/customer.action.ts:
```diff
@@ -44,9 +56,41 @@ export class CustomerSearchReset extends StateUtils.LoaderResetAction {
   }
 }
 
+export class CustomerListCustomersSearch extends StateUtils.LoaderLoadAction {
```
This action.ts looks good me, just one small thing: Why `CustomerListCustomersSearch`? (there are 2 customer in the name). how about `CustomerListSearch`?

WeizhengSap commented on feature-libs/asm/core/facade/facade-providers.ts:
```diff
@@ -0,0 +1,17 @@
+  AsmCustomerListService,
+  {
+    provide: AsmCustomerListFacade,
+    useExisting: AsmCustomerListService,
```
Do we want to make `AsmService` as facade?

ercultimate replied:
`AsmService` cannot be converted to a facade without breaking changes.

WeizhengSap commented on feature-libs/asm/core/facade/asm.service.ts:
```diff
@@ -20,7 +22,24 @@ import { AsmSelectors } from '../store/index';
+    @Optional() protected asmFacade?: AsmCustomerListFacade
+  ) {}
+
+  getCustomerLists(): Observable<QueryState<CustomerListsPage>> | undefined {
```
since we have `AsmCustomerListFacade`, so no need add this function in `AsmService`, right? Customer can get this function from `AsmCustomerListFacade`.

WeizhengSap commented on angular.json:
```diff
@@ -1,5 +1,8 @@
   "$schema": "./node_modules/@angular/cli/lib/config/schema.json",
+  "cli": {
+    "analytics": false
+  },
```
can we revert it back?

ercultimate committed:
remove local angular cli settings

ercultimate committed:
move getCustomerLists to customer list facade

ercultimate committed:
normaize errors for customer list request

WeizhengSap commented on feature-libs/asm/occ/adapters/occ-asm.adapter.ts:
```diff
@@ -35,27 +39,70 @@ export class OccAsmAdapter implements AsmAdapter {
+
+    return this.http.get<CustomerListsPage>(url, { headers, params }).pipe(
+      this.converterService.pipeable(CUSTOMER_LISTS_NORMALIZER),
+      catchError((error) => throwError(normalizeHttpError(error)))
```
should catchError before convert, right?

ercultimate replied:
I don't believe the order really matters for these operations, but I see that conventionally the error normalization is done first. 

The `pipeable` method uses `map` and does not interact with the observable's error out, so an error would pass by the converter with either ordering.

WeizhengSap commented on feature-libs/asm/core/facade/asm.service.ts:
```diff
@@ -53,6 +53,38 @@ export class AsmService {
+  /**
+   * Search for customers in a customer list
+   */
+  customerListCustomersSearch(options: CustomerSearchOptions): void {
```
can we also move them to `AsmCustomerListService`?

ercultimate replied:
Not at the moment.  The models used are located in the `core` module while the facade is located in the `root` module.  Moving the methods would require a breaking change to move the models to `root`.

WeizhengSap commented on feature-libs/asm/root/facade/asm-customer-list.facade.ts:
```diff
@@ -0,0 +1,25 @@
+})
+export abstract class AsmCustomerListFacade {
+  abstract getCustomerLists(): Observable<CustomerListsPage | undefined>;
+  abstract getCustomerListsState(): Observable<QueryState<CustomerListsPage>>;
```
this one is not used, right? I am fine to expose it.

ercultimate replied:
Correct, it is not used.  I had seen other services that use queries exposing both.

WeizhengSap commented on feature-libs/asm/core/facade/asm-customer-list.service.ts:
```diff
@@ -0,0 +1,33 @@
+export class AsmCustomerListService implements AsmCustomerListFacade {
+  protected customerListQuery$: Query<CustomerListsPage> =
+    this.queryService.create(() => this.asmConnector.customerLists(), {
+      reloadOn: undefined,
```
do we want to reload the data when language change?

ercultimate committed:
reorder adapter conversion and error normalization to match convention

ercultimate committed:
Merge remote-tracking branch 'origin/develop' into feature/cxspa-206

ercultimate committed:
fix duplicate facadeProvider spread from merge

ercultimate committed:
clean up development code notes

WeizhengSap commented on projects/storefrontlib/cms-components/misc/icon/icon.model.ts:
```diff
@@ -45,8 +45,14 @@ export enum ICON_TYPE {
   OFF = 'OFF',
   LINK_OUT = 'LINK_OUT',
   EXPAND_ARROWS = 'EXPAND_ARROWS',
+  CLIPBOARD_LIST = 'CLIPBOARD_LIST',
```
If adding them is breaking change, maybe augment this enum in asm.

ercultimate replied:
I don't believe this is a breaking change.  We are only adding new values to the enum, so no previously used values would be changed.  Additionally, enums cannot be extended, so there is little chance of conflicts.

WeizhengSap commented on projects/storefrontlib/cms-components/misc/icon/fontawesome-icon.config.ts:
```diff
@@ -45,9 +45,15 @@ export const fontawesomeIconConfig: IconConfig = {
       SORT_DOWN: 'fas fa-sort-down',
       SORT: 'fas fa-sort',
       EXPAND_ARROWS: 'fas fa-expand',
+      CLIPBOARD_LIST: 'fas fa-clipboard-list',
```
If adding them is breaking change, maybe augment this config in asm.

ercultimate replied:
I don't believe this is a breaking change.  We are only adding new values to the configuration, so no previously used values would be changed.  The configurations are deep merged into one another, so any conflicting values provided by the merchant would be overridden.

WeizhengSap commented on feature-libs/asm/core/connectors/asm.connector.spec.ts:
```diff
@@ -64,13 +85,27 @@ describe('AsmConnector', () => {
+    asmConnector.customerLists().subscribe((results) => {
+      expect(results).toEqual(mockCustomerListPage);
+      done();
+    });
```
missing test for bindCart

ercultimate replied:
Tests added.

ercultimate committed:
add unit tests for asm connector bind cart method

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

hakwookim committed:
replaced ModalService with LaunchDialogService

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

hakwookim committed:
Merge remote-tracking branch 'origin/develop' into feature/cxspa-206

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

WeizhengSap commented on feature-libs/asm/root/config/asm-config.ts:
```diff
@@ -3,7 +3,9 @@
+import { ICON_TYPE } from '@spartacus/storefront';
+import { CustomerListColumnActionType } from '../model/customer-list.model';
 
 export abstract class AsmConfig extends OccConfig {
```
no need to `extends OccConfig`, right? This config has nothing related to occ.

justinlee01 replied:
I suppose so, if it is old. `CmsConfig` and `AnonymousConsentsConfig` extend `OccConfig`, unless they're not meant to.

WeizhengSap replied:
CmsConfig needs extends OccConfig, but I don't think AnonymousConsentsConfig should extend OccConfig.

kimhw0630 replied:
extends OccConfig removed.. thanks

justinlee01 commented on feature-libs/asm/components/customer-list/customer-list.component.html:
```diff
@@ -0,0 +1,354 @@
+</ng-template>
+<ng-template #cellContent let-customerEntry="customerEntry" let-column="column">
+  <span *ngIf="!column.icon">{{
+    column.renderer ? column.renderer(customerEntry) : ''
```
I want to say Angular supports the Elvis operator in template HTML now
```suggestion
    column.renderer?.(customerEntry) || ''
```

WeizhengSap commented on feature-libs/asm/root/config/asm-config.ts:
```diff
@@ -13,6 +15,19 @@ export abstract class AsmConfig extends OccConfig {
     customerSearch?: {
       maxResults?: number;
     };
+    customerList?: {
```
Why some of the asm config is moved from core to root?

justinlee01 replied:
I can explain this.
For this PR (https://github.com/SAP/spartacus/pull/16095) I added an interceptor `UserIdHttpHeaderInterceptor` that lived in `@spartacus/asm/root`.
As it needed to be turned on/off - depending on the customer's backend - a configuration was added to `AsmConfig`.
This caused a circular dependency - `@spartacus/asm/root` depended on `@spartacus/asm/core` which depends on `root`.
Patrick and I concluded the configuration should have been in `root` originally. There is precedence for this - for example, `CartConfig` is in root.
Let me know if you have any further questions.

WeizhengSap commented on feature-libs/asm/components/asm-main-ui/asm-main-ui.component.ts:
```diff
@@ -160,32 +174,32 @@ export class AsmMainUiComponent implements OnInit {
+    );
+
+    this.subscription.add(
+      this.launchDialogService?.dialogClose
```
each time function `showCustomList` is called, a new subscription will be created. So, the code block `    this.subscription.add(...)` can be moved into `ngOnInit`, right?

kimhw0630 replied:
thanks, you are right we can move into ngOnInit.

hakwookim committed:
code review changes

hakwookim committed:
code review: removed extend OccConfig

kimhw0630 committed:
Merge branch 'develop' into feature/cxspa-206

justinlee01 commented on feature-libs/asm/components/customer-list/customer-list.component.ts:
```diff
@@ -0,0 +1,264 @@
+    action: CustomerListColumnActionType
+  ): void {
+    this.selectedCustomer = customerEntry;
+    let closeValue: CustomerListAction = {
```
```suggestion
    const closeValue: CustomerListAction = {
```

kimhw0630 replied:
we will update in ASM 360

justinlee01 commented on feature-libs/asm/components/customer-list/customer-list.component.ts:
```diff
@@ -0,0 +1,264 @@
+  goToNextPage(): void {
+    if (this.currentPage >= this.maxPage) {
+      this.currentPage = this.maxPage;
+    } else {
+      if (this.loaded) {
+        this.currentPage++;
+        this.fetchCustomers();
+      }
+    }
```
```suggestion
    } else if (this.loaded) {
      this.currentPage++;
      this.fetchCustomers();
    }
```

kimhw0630 replied:
we will update in ASM 360

justinlee01 commented on feature-libs/asm/components/customer-list/customer-list.component.ts:
```diff
@@ -0,0 +1,264 @@
+          textByOrderDateDesc,
+          textByDateAsc,
+          textByDateDesc,
+        ]) => {
+          return {
+            byNameAsc: textByNameAsc,
+            byNameDesc: textByNameDesc,
+            byOrderDateAsc: textByOrderDateAsc,
+            byOrderDateDesc: textByOrderDateDesc,
+            byDateAsc: textByDateAsc,
+            byDateDesc: textByDateDesc,
+          };
+        }
```
I'm surprised prettier did not truncate ths

```suggestion
        ]) => ({
          byNameAsc: textByNameAsc,
          byNameDesc: textByNameDesc,
          byOrderDateAsc: textByOrderDateAsc,
          byOrderDateDesc: textByOrderDateDesc,
          byDateAsc: textByDateAsc,
          byDateDesc: textByDateDesc,
        })
```

kimhw0630 replied:
we will update in ASM 360
